File: backend/app/schemas/camera.py
from __future__ import annotations

from datetime import datetime

from pydantic import BaseModel, Field


class CameraCreate(BaseModel):
    name: str
    classroom_id: str
    rtsp_host: str
    rtsp_port: int = 554
    rtsp_path: str = ""
    rtsp_username: str = ""
    rtsp_password: str = Field(default="", exclude=True)  # never echoed back
    fps: int = 25
    resolution_width: int = 1920
    resolution_height: int = 1080
    enabled: bool = True


class CameraUpdate(BaseModel):
    name: str | None = None
    rtsp_host: str | None = None
    rtsp_port: int | None = None
    rtsp_path: str | None = None
    rtsp_username: str | None = None
    rtsp_password: str | None = Field(default=None, exclude=True)
    fps: int | None = None
    resolution_width: int | None = None
    resolution_height: int | None = None
    enabled: bool | None = None


class CameraOut(BaseModel):
    """Never includes rtsp_password / rtsp_password_encrypted — credentials
    are write-only from the API's perspective (spec §18: do not expose RTSP
    credentials in frontend responses)."""

    id: str
    name: str
    classroom_id: str
    rtsp_host: str
    rtsp_port: int
    rtsp_path: str
    rtsp_username: str
    fps: int
    resolution_width: int
    resolution_height: int
    status: str
    enabled: bool
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class ROIPoint(BaseModel):
    x: float
    y: float


class CalibrationPoint(BaseModel):
    pixel_y: float
    reference_height_px: float


class CameraROIUpdate(BaseModel):
    roi: list[ROIPoint]


class CameraCalibrationUpdate(BaseModel):
    calibration_points: list[CalibrationPoint]
    adult_height_ratio: float | None = None
    child_height_ratio: float | None = None


class CameraConfigurationOut(BaseModel):
    camera_id: str
    roi: list[ROIPoint]
    calibration_points: list[CalibrationPoint]
    adult_height_ratio: float
    child_height_ratio: float

File: frontend_legacy_ts_tailwind/src/types/camera.ts
export type CameraStatus = 'ONLINE' | 'OFFLINE' | 'DISABLED';

export interface Camera {
  id: string;
  name: string;
  classroom_id: string;
  rtsp_host: string;
  rtsp_port: number;
  rtsp_path: string;
  rtsp_username: string;
  fps: number;
  resolution_width: number;
  resolution_height: number;
  status: CameraStatus;
  enabled: boolean;
  created_at: string;
  updated_at: string;
}

/** Body for POST /api/cameras. Password is write-only and never echoed back by the API. */
export interface CameraCreateInput {
  name: string;
  classroom_id: string;
  rtsp_host: string;
  rtsp_port: number;
  rtsp_path: string;
  rtsp_username: string;
  rtsp_password: string;
  fps: number;
  resolution_width: number;
  resolution_height: number;
  enabled: boolean;
}

/** Body for PUT /api/cameras/{id}. Every field optional; omit rtsp_password to leave it unchanged. */
export type CameraUpdateInput = Partial<CameraCreateInput>;

export interface CameraTestResult {
  success: boolean;
  message: string;
}

export interface RoiPoint {
  x: number;
  y: number;
}

export interface CalibrationPoint {
  pixel_y: number;
  reference_height_px: number;
}

export interface CameraConfiguration {
  camera_id: string;
  roi: RoiPoint[];
  calibration_points: CalibrationPoint[];
  adult_height_ratio: number;
  child_height_ratio: number;
}

export interface RoiUpdateInput {
  roi: RoiPoint[];
}

export interface CalibrationUpdateInput {
  calibration_points: CalibrationPoint[];
  adult_height_ratio?: number;
  child_height_ratio?: number;
}
